macaroon_service_test.go

```diff
@@ -94,6 +100,159 @@ func TestMacaroonServiceMigration(t *testing.T) {
 	require.NoError(t, ms.Stop())
 }
 
+// TestMacaroonGeneration tests that the macaroon service generates macaroons
+// as expected. A macaroon should not be regenerated if the required permissions
+// haven't changed, and should be regenerated if the required permissions have
+// changed.
+func TestMacaroonGeneration(t *testing.T) {
+	// Create a temporary directory where we can store the macaroon db and
+	// the macaroon we are about to create.
+	tempDirPath := t.TempDir()
+	macaroonPath := filepath.Join(tempDirPath, "test.macaroon")
+
+	// Set the macaroon store, and get the root key store.
+	rks, backend, err := NewBoltMacaroonStore(
+		tempDirPath, "macaroons.db", defaultMacaroonTimeout,
+	)
+	require.NoError(t, err)
+	t.Cleanup(func() {
+		require.NoError(t, backend.Close())
+	})
+
+	// Create the initial config. We set a mock DB password, as we either
+	// need to set a DB password or provide an LndClient, ephemeral key and
+	// key locator so that a shared key can be derived with LND, when
+	// creating the macaroon service.
+	cfg := &MacaroonServiceConfig{
+		MacaroonLocation: tempDirPath,
+		DBPassword:       []byte{1, 1, 1, 1, 1, 1, 1, 1, 1, 1},
+		RootKeyStore:     rks,
+		MacaroonPath:     macaroonPath,
+	}
+
+	// Generate mock permissions.
+	firstEntityWrite := bakery.Op{Entity: "first_entity", Action: "write"}
+	firstEntityRead := bakery.Op{Entity: "first_entity", Action: "read"}
+	secondEntityRead := bakery.Op{Entity: "second_entity", Action: "read"}
+
+	// Firstly, we test that will fail to create a macaroon if we don't
+	// set cfg.RequiredPerms, or set the permissions to an empty map.
+	ms, err := NewMacaroonService(cfg)
+	require.NoError(t, err)
+
+	// We start the macaroon service, which will fail as we haven't set the
+	// required permissions.
+	require.Error(t, ms.Start())
+	require.NoError(t, ms.Stop())
+
+	// We now set the required permissions to an empty map, and ensure that
+	// the macaroon service fails to start.
+	mockPerms := map[string][]bakery.Op{}
+	require.Error(t, ms.Start())
+	require.NoError(t, ms.Stop())
+
+	// We add 2 required permissions to the mockPerms map, which would make
+	// the previous start attempts succeed if we start the service now.
+	mockPerms["/uri1.Test/WriteTest"] = []bakery.Op{firstEntityWrite}
+	mockPerms["/uri1.Test/ReadTest"] = []bakery.Op{firstEntityRead}
+	cfg.RequiredPerms = mockPerms
+
+	// But we also test that we fail to start the macaroon service if we set
+	// the MacaroonPath to an existing file which isn't a macaroon.
+	cfg.MacaroonPath = filepath.Join(tempDirPath, "macaroons.db")
```

Comment:
Just wanted to check:
If we set the macaroon path a non existing file with another file extension than .macaroon, we'll still generate the macaroon, and write the file to that path.
This is expected behavior correct, as it's technically a macaroon file even though it has another file extension?

Reply:
yeah I dont think the extension matters :) I think the behaviour is expected :)

Reply:
Awesome thanks!